### DISCUSSION TITLE
How to respond to WebSocket Ping message?

### DISCUSSION THREAD

**Discussion by otsuka** [fastapi/fastapi#7940] [Questions] [answered]
When a WebSocket Ping message (data frame) is sent to a FastAPI WebSocket server, how can I receive it and return a Pong message?

Sending Ping using aiohttp as below, it doesn't seem that FastAPI WebSocket connection receives anything.

###### Sending Ping

```py
async with aiohttp.ClientSession() as sess:
    async with sess.ws_connect(WS_ENDPOINT) as ws:
        await ws.ping()
```

###### Receiving Ping

```py
@app.websocket("/ws")
async def websocket_endpoint(ws: WebSocket):
    await ws.accept()
    data = await ws.receive()
```

**Comment by euri10** [comment]
~~imho the simplest is not to use FastAPI decorator but subclassing Starlette `WebSocketEndpoint` and sending your pong in the `on_receive` method~~ 

Oo that ping... :smile:

**Comment by tiangolo** [comment, marked as answer]
Ping events are defined in the WebSocket spec: https://tools.ietf.org/html/rfc6455#section-5.5.2 and are handled automatically by Uvicorn, not by FastAPI.

**Comment by otsuka** [comment]
@tiangolo 
Thank you for your response.
I found the code that handles a Ping message in Uvicorn.
https://github.com/encode/uvicorn/blob/master/uvicorn/protocols/websockets/wsproto_impl.py#L186-L187

**Reply by markus-work** [reply]
Here is a permalink for those who see this later: https://github.com/encode/uvicorn/blob/14ffba8316eb606cd026d1a3b01d9d90e47e868c/uvicorn/protocols/websockets/wsproto_impl.py#L218-L219

**Comment by BoilingNine** [comment]
```python
uvicorn.run("main:app",
                host=settings.WEB_HOST,
                port=settings.WEB_PORT,
                ws_ping_interval=20,ws_ping_timeout=20)
``` 
I did it but timeout not work,I don't know how to solve it
